src/hotspot/os/bsd/memMapPrinter_macosx.cpp

```diff
@@ -0,0 +1,375 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#if defined(__APPLE__)
+
+#include "precompiled.hpp"
+
+#include "nmt/memMapPrinter.hpp"
+#include "runtime/os.hpp"
+#include "utilities/align.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/powerOfTwo.hpp"
+
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <libproc.h>
+#include <unistd.h>
+
+#include <mach/vm_inherit.h>
+#include <mach/vm_prot.h>
+#include <mach/mach_vm.h>
+
+/* maximum number of mapping records returned */
+static const int MAX_REGIONS_RETURNED = 1000000;
+
+class MappingInfo {
+public:
+  const char* _address;
+  size_t _size;
+  stringStream _share_buffer;
+  stringStream _type_buffer;
+  stringStream _protect_buffer;
+  stringStream _file_name;
+  const char* _tag_text;
+
+  MappingInfo() {}
+
+  void process(const proc_regionwithpathinfo& mem_info) {
+    _share_buffer.reset();
+    _protect_buffer.reset();
+    _type_buffer.reset();
+    _file_name.reset();
+    _tag_text = "";
+
+    const proc_regioninfo& rinfo = mem_info.prp_prinfo;
+
+    _address = (const char*) rinfo.pri_address;
+    _size = rinfo.pri_size;
+
+    const char* path = mem_info.prp_vip.vip_path;
+    if (path != nullptr) {
+      _file_name.print_raw(path);
+    } else {
+      char buf[PATH_MAX];
+      buf[0] = 0;
+      proc_regionfilename(getpid(), (uint64_t) _address, buf, sizeof(buf));
+      if (buf[0] != 0) {
+        _file_name.print_raw("-> ");
+        _file_name.print_raw(buf);
+      }
+    }
+
+    char prot[4];
+    rwbits(rinfo.pri_protection, prot);
+    _protect_buffer.print("%s", prot);
+
+    get_share_mode(_share_buffer, rinfo);
+    _tag_text = tagToStr(rinfo.pri_user_tag);
+  }
+
+  static void get_share_mode(outputStream& out, const proc_regioninfo& rinfo) {
+    static const char* share_strings[] = {
+      "cow", "pvt", "---", "shr", "tsh", "p/a", "s/a", "lpg"
+    };
+    const bool valid_share_mode = rinfo.pri_share_mode >= SM_COW && rinfo.pri_share_mode <= SM_LARGE_PAGE;
+    if (valid_share_mode) {
+      int share_mode = rinfo.pri_share_mode;
+      //share_mode = share_mode == SM_LARGE_PAGE || share_mode == SM_PRIVATE_ALIASED ? SM_PRIVATE;
+      out.print_raw(share_strings[share_mode - 1]);
+    } else {
+      out.print_cr("invalid pri_share_mode (%d)", rinfo.pri_share_mode);
+      assert(valid_share_mode, "invalid pri_share_mode (%d)", rinfo.pri_share_mode);
+    }
+    if (rinfo.pri_flags & PROC_REGION_SHARED) {
+        out.print_raw("-shared");
+    }
+    if (rinfo.pri_flags & PROC_REGION_SUBMAP) {
+        out.print_raw("-submap");
+    }
+    if ((rinfo.pri_flags & (PROC_REGION_SHARED | PROC_REGION_SUBMAP)) != rinfo.pri_flags) {
+      out.print_cr("unhandled pri_flags = 0x%x", rinfo.pri_flags);
+      assert(false, "unhandled pri_flags = 0x%x", rinfo.pri_flags);
+    }
+  }
+
+  static const char* tagToStr(uint32_t user_tag) {
+    switch (user_tag) {
```

Comment:
You could reduce the code and remove duplications with an [x macro](https://en.wikipedia.org/wiki/X_macro) if you wanted.

Reply:
Doing this, I found a copy-paste error, so thanks for the suggestion!